Add regional evolution metadata support

joereg4 opened the pull request:
- Add region_restriction and base_form_required fields to PokemonEvolution model
- Update PokemonEvolutionSerializer to include new fields in API responses
- Add migration for new fields
- Add implementation plan documentation
- Add data import script for regional evolution data

This addresses GitHub issue #639 by adding support for regional evolution requirements like Galarian Yamask → Runerigus and Galarian Slowpoke → Slowbro.

pauljoda commented:
I think this is a no brainer, as currently the api has a huge oversight as far as how this is handled. 

This references a pokemon species for the form foreign key, it's my understanding that the species is the reference to the base forms data, though that object does contain a list of all forms. Due to this needing to reference a specific form, would it make more sense to instead link to the Pokémon id itself, and if species data is needed it can be queried from that Pokémon's species? This ensures the link is to the exact form required for the evolution, and does not require more implied logic to parse out based on the name passed with the species link.

Naramsim commented:
I'm away for some days and can't review these PRs at the moment. Thanks for submitting it

Naramsim commented:
Hi! Can you fix the CICD pipeline and continue your work with AI? It's promising.

joereg4 commented:
@Naramsim fixed the CI issues:
- Updated migration dependency to reference correct parent migration
- Applied black formatting to match project standards
- Removed orphaned database table causing test failures

All CI checks are now passing.

joereg4 commented:
@Naramsim Good morning, not sure if there is anything I can do regarding the graphql load timing out in the Container / k8s check.  If there is something i can tweak please let me know:)

joereg4 commented:
@Naramsim Seeing that the loading of graphql is timing out.  Added some debugging to see where it's getting hung up and also increased the timeout in case the CI environment is just slow.

joereg4 commented:
@Naramsim I've been working the load-graphql issues and was able to resolve it.  I see that the last dependabot commit (ee8ef9ffb24a61b6943bf2bb88fb3c9b1d8e5150) failed for the same reason.  debian:buster was end of life and a switch to debian:bullseye in the load-graphql.yaml fixed the issue. 

Once this PR is approved the workflows should pass.  Was able to test with github actions on my fork.

joereg4 commented on import_regional_evolutions.py:
```diff
@@ -0,0 +1,162 @@
+#!/usr/bin/env python3
+"""
```
Made the change using the build process

pauljoda commented:
I can take a crack at it, for reference I'm using the data for an iOS app, and actually used the Postgres dump to load it into a sqlite db to ship with the app, it quickly got way too network heavy to use the api calls. 

This would be a pretty big refactor, are you saying to flatten all forms, species, and varieties into the Pokémon table itself? I think there is some benefit to having a way for forms to share a common data with the root Pokémon, providing the form is mostly cosmetic. I do think that regional forms should be completely separate Pokémon. 

Perhaps, the structure of Pokémon/Species could still be used, but with the purpose now being Form/Pokemon, where the species table now takes the role of the base pokemon, and the pokemon table now takes the role of the differentiating factors per form. 

Any thoughts or suggestions before I work on that? I think that would be a far more future proof and simple method than adding a new table here, though this pr does solve the problem with the current API so might be worth pulling into for a quick fix and this change come in a future version.

pauljoda commented:
> @joereg4 I'm sure it's fine, I think large changes like this should be for v3. I'll move this discussion to a thread so this can remain related for this change on v2

https://github.com/PokeAPI/pokeapi/discussions/1305

Naramsim commented on data/v2/csv/pokemon_evolution.csv:
```diff
@@ -1,508 +1,508 @@
-509,997,1,,35,,,,,,,,,,,,,,0,0
-510,998,1,,54,,,,,,,,,,,,,,0,0
-511,1000,10,,,,,,,,,,,,,,,,0,0
+id,evolved_species_id,evolution_trigger_id,trigger_item_id,minimum_level,gender_id,location_id,held_item_id,time_of_day,known_move_id,known_move_type_id,minimum_happiness,minimum_beauty,minimum_affection,relative_physical_stats,party_species_id,party_type_id,trade_species_id,needs_overworld_rain,turn_upside_down,region_restriction,base_form_required
```
Great job! We are close to merging this in! But beforehand I'd like to clean a bit the names used. Can we change `region_restriction` to `region_id` and `base_form_required` to `base_form` or `base_form_id` (if it's an ID)? So that the colum names are inline with the existing fields, such as `region_id`.

joereg4 replied:
Done, pushed the change

Naramsim commented on pokemon_v2/models.py:
```diff
@@ -1694,6 +1694,24 @@ class PokemonEvolution(HasEvolutionTrigger, HasGender):
     turn_upside_down = models.BooleanField(default=False)
 
+    # Regional evolution fields
+    region_restriction = models.ForeignKey(
```
If you agree I'd go with the same names here.

joereg4 replied:
Done, pushed the change

joereg4 commented:
@Naramsim Made the changes requested, please take a look:)

pokeapi-machine-user commented:
A [PokeAPI/api-data](https://github.com/PokeAPI/api-data) refresh has started. In ~45 minutes the staging branch of [PokeAPI/api-data](https://github.com/PokeAPI/api-data/tree/staging) will be pushed with the new generated data. <br><br> The staging branch will be deployed in our [staging environment](https://staging.pokeapi.co/api/v2/) and the entire API will be ready to review. <br><br> A Pull Request ([master](https://github.com/PokeAPI/api-data/tree/master)<-[staging](https://github.com/PokeAPI/api-data/tree/staging)) will be also created at [PokeAPI/api-data](https://github.com/PokeAPI/api-data/pulls) and assigned to the PokeAPI Core team to be reviewed. If approved and merged new data will soon be available worldwide at [pokeapi.co](https://pokeapi.co/api/v2/).

Naramsim commented on data/v2/csv/pokemon_evolution.csv:
```diff
@@ -1,508 +1,508 @@
+10,15,1,,10,,,,,,,,,,,,,,0,0,,
+11,17,1,,18,,,,,,,,,,,,,,0,0,,
+12,18,1,,36,,,,,,,,,,,,,,0,0,,
+13,20,1,,20,,,,,,,,,,,,,,0,0,7,19
```
@joereg4 could you take a look at the above?

joereg4 replied:
The evolution data for Rattata → Raticate was incorrectly restricted to the Alola region only, preventing the evolution from occurring in other regions. This was identified by @Naramsim in the review of the evolution data.

Issue Details:

- Evolution entries had region_id=7 (Alola) and base_form_id=19 (Rattata)
- This incorrectly suggested that Rattata could only evolve into Raticate in Alola
- The evolution should be available in all regions, not just Alola

#1308
